src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs

```diff
@@ -601,6 +677,46 @@ public ProjectState AddAdditionalDocument(TextDocumentState document)
                 additionalDocumentStates: _additionalDocumentStates.Add(document.Id, document));
         }
 
+        public ProjectState AddAnalyzerConfigDocuments(ImmutableArray<AnalyzerConfigDocumentState> documents)
+        {
+            Debug.Assert(!documents.Any(d => this._analyzerConfigDocumentStates.ContainsKey(d.Id)));
+
+            var newAnalyzerConfigDocumentStates = _analyzerConfigDocumentStates.AddRange(documents.Select(d => KeyValuePairUtil.Create(d.Id, d)));
+
+            return CreateNewStateForChangedAnalyzerConfigDocuments(newAnalyzerConfigDocumentStates);
+        }
+
+        public ProjectState RemoveAnalyzerConfigDocument(DocumentId documentId)
+        {
+            Debug.Assert(_analyzerConfigDocumentStates.ContainsKey(documentId));
+
+            var newAnalyzerConfigDocumentStates = _analyzerConfigDocumentStates.Remove(documentId);
+
+            return CreateNewStateForChangedAnalyzerConfigDocuments(newAnalyzerConfigDocumentStates);
+        }
+
+        private ProjectState CreateNewStateForChangedAnalyzerConfigDocuments(ImmutableSortedDictionary<DocumentId, AnalyzerConfigDocumentState> newAnalyzerConfigDocumentStates)
+        {
+            var newAnalyzerConfigSet = ComputeAnalyzerConfigSetValueSource(newAnalyzerConfigDocumentStates.Values);
+
+            // The addition of any .editorconfig can modify the diagnostic reporting options that are on
+            // a specific syntax tree; therefore we must update all our syntax trees.
+            var docMap = _documentStates;
+
+            foreach (var (docId, _) in _documentStates)
+            {
+                var oldDocState = this.GetDocumentState(docId);
+                var newDocState = oldDocState.UpdateAnalyzerConfigSet(newAnalyzerConfigSet);
+                docMap = docMap.SetItem(docId, newDocState);
+            }
+
+            return this.With(
+                projectInfo: this.ProjectInfo.WithVersion(this.Version.GetNewerVersion()),
+                analyzerConfigDocumentStates: newAnalyzerConfigDocumentStates,
+                documentStates: docMap,
+                analyzerConfigSet: newAnalyzerConfigSet);
+        }
+
         public ProjectState RemoveDocument(DocumentId documentId)
```

Comment:
i'm on the fence, but maybe as an organizational tool, you culd move tehse to a partial part.

that said, i see the desire to keep these organized with the other simialr methods.

Reply:
also on the fence about all the new methods being placed between the existng add/remove methods.

Reply:
The existing organization in the file was to put the add methods together, then the remove methods. I agree that perhaps organizing by 'noun' instead of 'verb' makes more sense. Happy to take a follow up item to reorganize.